Cylindrical transformations with well-defined phi angle

jngrad opened the pull request:
Follow-up to #4014

Description of changes:
- add overloads to convert from Cartesian to cylindrical coordinates with custom longitudinal axis and phi angle

jngrad committed:
utils: Add modulo function for radians

jngrad committed:
tests: Add tolerance for floating point operations

jngrad committed:
utils: Refactor coordinate transformation code

Provides 3 overloads to transform between Cartesian and:
- the cylindrical system
- a cylindrical system with custom axis (+axis parameter)
- an oriented cylindrical system with custom axis (+phi0 parameter)

jngrad committed:
utils: Fix regressions

KaiSzuttor commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,125 @@
+ * with change of basis.
+ * @param pos    %Vector to transform
+ * @param axis   Longitudinal axis of the cylindrical coordinates
+ * @param phi0   Reference angle for @f$ \phi = 0 @f$
```
I'm not sure if this is clear looking at the function. You should explain what you mean by "Reference angle"

KaiSzuttor commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,125 @@
+ * with change of basis.
+ * @param pos    %Vector to transform
+ * @param axis   Longitudinal axis of the cylindrical coordinates
+ * @param phi0   Reference point at which @f$ \phi = 0 @f$
```
this should probably be called ```orientation``` or similar.

KaiSzuttor replied:
is this reference point given in the untransformed or the transformed coordinate system?

KaiSzuttor commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,125 @@
+ * @brief Coordinate transformation from cylindrical to Cartesian coordinates.
+ * @param pos    %Vector to transform
+ * @param axis   Longitudinal axis of the cylindrical coordinates
+ * @param phi0   Reference point at which @f$ \phi = 0 @f$
```
```orientation``` instead of ```phi0```?

KaiSzuttor commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,125 @@
+ */
+inline Vector3d transform_coordinate_cartesian_to_cylinder(
+    Vector3d const &pos, Vector3d const &axis, Vector3d const &phi0) {
+  auto const phi = transform_coordinate_cartesian_to_cylinder(phi0, axis)[1];
```
i think it would be cleaner if we had a totally independent function for this:
``` c++
double angle_between(Utils::Vector3d const &a, Utils::Vector3d const &b) {
  return a*b / (a.norm() * b.norm());
}
```
```

jngrad committed:
utils: Simplify transformation code

Remove the Utils::rotation_params() function in favor of a more basic
utility function: Utils::angle_between(). This makes clear at the
call sites what is actually calculated.

jngrad committed:
utils: Fix doxygen documentation

jngrad committed:
tests: Remove duplicated checks

RudolfWeeber commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,133 @@
+/**
+ * @brief Coordinate transformation from Cartesian to cylindrical coordinates.
+ * @param pos    %Vector to transform
  */
```
Please mention that that the origins and z-axes of the coordinate systems co-incide and that the phi=0 direction corresponds to the x axis in the original coordinate system.

RudolfWeeber commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,133 @@
+ * with change of basis.
+ * @param pos    %Vector to transform
+ * @param axis   Longitudinal axis of the cylindrical coordinates
+ * @param phi0   Value for which @f$ \phi = 0 @f$, is obtained by the
```
Please write down the meaning of phi as an equation. I don't really understand this function.

RudolfWeeber commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,133 @@
+}
+
+/**
+ * @brief Coordinate transformation from Cartesian to cylindrical coordinates
```
Please mention that the origins of the ccordinate systems co-incide

RudolfWeeber commented on src/utils/include/utils/math/coordinate_transformation.hpp:
```diff
@@ -19,57 +19,133 @@
+ * @brief Coordinate transformation from cylindrical to Cartesian coordinates.
+ * @param pos    %Vector to transform
+ * @param axis   Longitudinal axis of the cylindrical coordinates
+ * @param phi0   Value for which @f$ \phi = 0 @f$, is obtained by the
```
Again, meaing of phi0 is not clear to me. For the back-transofmrs please also mention, what co-incides and what does not, for each case

RudolfWeeber commented on src/utils/tests/coordinate_transformation.cpp:
```diff
@@ -80,8 +151,58 @@ BOOST_AUTO_TEST_CASE(cylinder_to_cartesian_test) {
+      BOOST_CHECK_SMALL(abs(u_cart[i] - u[i]), eps);
+      BOOST_CHECK_SMALL(abs(v_cart[i] - v[i]), eps);
+    }
   }
```
Please add a test for right-handedness:
* let v1 and v2 be reasonable random vectors
* use a = vector_rpoduct(v1,v2) / |v1|/|v2|
* transform v1 with axis a and orientation v1, assert phi=0, r=|v1|,z=0
* transform v2 with axis a and orientation v1, assert phi=acos(<v1,v2>/|v1|/|v2|), r=|v1|,z=0
* transform v1 with axis a and orientation v2, assert phi=-1 *acos(<v1,v2>/|v1|/|v2|), r=|v1|,z=0

or something similar.

fweik replied:
```c++
bool is_right_handed = vector_product(v1, v2) > 0;
```
?

jngrad replied:
done in 51830528

RudolfWeeber reviewed (changes requested):
I have a hard time understanding the approach with the `phi_0`.

I'd hae preferred the approach with a basis transform, analogous to Python code below, since, at least for me, it is clearer how and that that works.

If you'd like to stay with the current implementation, please add an explanation in the comments explaining the approach, particularly the role of phi_0. 

Could you please also consider the tests from `test_to_cylindrical_basis_change` from the below Python code. They are constructed such that the expected results can be inferred easily by a human.

RudolfWeeber commented:
``` python
import numpy as np

def basis_change(b1,b2,b3, v):
    M = np.linalg.inv(np.array((b1/np.linalg.norm(b1),b2/np.linalg.norm(b2),b3/np.linalg.norm(b3))).T)
    return np.dot(M, v)

def to_cylindrical(x):
    return (np.sqrt(x[0]**2 +x[1]**2), np.arctan2(x[1], x[0]), x[2])


def to_cylindrical_basis_change(x, axis=None, orientation=None):
    x_t=basis_change(orientation, np.cross(axis,orientation), axis, x)
    return to_cylindrical(x_t)

def test_to_cylindrical():
    t = to_cylindrical(np.array((np.sqrt(4.5),np.sqrt(4.5),1.2)))
    np.testing.assert_allclose(t, [3,np.pi/4, 1.2])

b_x = np.array((1,0,0))
b_y = np.array((0,1,0))
b_z = np.array((0,0,1))

def test_basis_transform():
    v = [1,2,3]
   
    # identity transform
    np.testing.assert_allclose(basis_change(b_x,b_y,b_z,v), v)

    # swap coordinates
    np.testing.assert_allclose(basis_change(b_z,b_y,b_x,[v[2],v[1],v[0]]), v)
    
    v1 = np.array(([2,2,2])) /np.sqrt(12)
    v2 = np.array(([3,3,-6])) /np.sqrt(9+9+36)
    c = np.cross(v1,v2)
    v3 = c /np.linalg.norm(c)

    np.testing.assert_allclose(
        basis_change(v1,v2,v3, 0.1*v1+0.2*v2 -0.3*v3),
        [0.1,0.2,-0.3],atol=1E-10)


def test_to_cylindrical_basis_change():
    v1 = np.array((1,3,4))
    v2 = np.array((-3.0,7,2))
    axis = np.cross(v1,v2)
    axis /= np.linalg.norm(axis)


    np.testing.assert_allclose(
        to_cylindrical_basis_change(v1,axis=axis,orientation=v1),
        [np.linalg.norm(v1),0,0],atol=1E-10)
    
    angle_v1_v2 = np.arccos(np.dot(v1,v2)/np.linalg.norm(v1)/np.linalg.norm(v2))
    np.testing.assert_allclose(
        to_cylindrical_basis_change(v2+2*axis,axis=axis,orientation=v1),
        [np.linalg.norm(v2),angle_v1_v2,2],atol=1E-10)
    np.testing.assert_allclose(
        to_cylindrical_basis_change(v1+2*axis,axis=axis,orientation=v2),
        [np.linalg.norm(v1),-angle_v1_v2,2],atol=1E-10)
```

jngrad committed:
utils: Improve documentation

jngrad committed:
tests: Add cylindrical transformation test case

jngrad commented:
I've added the right-handedness test case and updated the documentation of the phi angle.

In the python code sample, I could not reproduce the following snippet in the C++ unit test:
```py
np.testing.assert_allclose(
        to_cylindrical_basis_change(v2+2*axis,axis=axis,orientation=v1),
        [np.linalg.norm(v2),angle_v1_v2,2],atol=1E-10)
```

When adding `2*axis` to the input vector `v2`, I get random values for `z` instead of `2`.

KaiSzuttor commented:
can this be closed because of #4114?
